docs/adrs/005-jobs.md

```diff
@@ -0,0 +1,285 @@
+
+```roc
+npm : Tool
+npm = systemTool "npm" 
+
+
+node : Tool
+node = systemTool "node"
+
+
+nodeModules : Job
+nodeModules =
+    job
+        {
+            command: exec npm [ "install" ],
+            inputs: [ "package.json", "package-lock.json" ],
+            outputs: [ "node_modules" ],
+        }
+
+
+uglifyjs : Tool
+uglifyjs
+    customTool
+        {
+            path: "\(nodeModules.outputs)/node_modules/.bin/uglifyjs",
+            tools: [ node ],
+            environment: {:
+                "NODE_PATH" => "\(nodeModules.outputs)/node_modules"
+            :}
+        }
+```
+
+- [ ] Surely there is a better way to do this.
+      Needs more thought.
+
+### CPU Hinting
+
+rbt uses job parameters to jobs to construct a build graph, which it will then walk in parallel wherever possible.
+This means that we can do a better job scheduling tasks when the tasks themselves are smaller—think compressing a single image instead of a whole directory of them.
+
+However, many compilers (like Zig, Rust/Cargo, Elm, Haskell, and Roc itself) have a monolithic compilation process where they take an entrypoint and manage compiling all the dependencies themselves.
+This is great both for programmers and compiler authors: it unlocks optimization opportunities and in many cases does away with the need for a separate build tool.
+
+Unfortunately, it's harder for a generic build tool like rbt to deal with those kinds of processes in a build step.
+The big question here: should we schedule work on other cores while the big process is running?
+We don't want to cause the CPU to do too much context switching!
+
+To fix this, jobs can give rbt enough information to do the right thing by specifying **CPU Hints**.
+A job can either say it takes a single core or that it will saturate all available cores.
+
+A future ADR (or just the implementation) will determine how we deal with CPU hints, but one possibility is to avoid starting new work when there's a saturating job running.
+Another would be to use some concept of CPU reservation to avoid starting a saturating job before it held the lock on all cores (of course, we'd have to be careful to avoid creating deadlocks here!)
+
+Anyway, here's how you might use this:
+
+```roc
+binary : Job
+binary =
+    job
+        {
+            tools: [ cargo ],
+            command: exec "cargo" [ "build", "--release" ],
+            outputs: [ "target/release/binary-name" ],
+            cpuHint: SaturatesCpus,
+        }
+```
+
+### Fixers
+
+Code formatters and some linters can automatically fix problems in source code.
+Jobs can specify a flag to make this explicitly OK.
+It might look like this:
+
+```roc
+elmFormat : Job
+elmFormat =
+    job
+        {
+            tools: [ elmFormat ],
+            command: exec "elm-format" [ "--yes" ],
+            inputs: [ "src/Main.elm", "src/OtherModule.elm" ],
+            modifiesInputs: True,
```

Comment:
🤔 I wonder if `Bool` might not be granular enough for `modifiesInputs` - e.g. for something like `eslint`, there's an input of the eslint config, which should *not* be modified, along with the source files...

one idea: `modifiesInputs` could take a tag union of the type `[ All, AllExcept (List Str), None ]`